### ISSUE TITLE
Ecto.Changeset.validate_required/2 should not silently pass nil as list of fields

### ISSUE THREAD

**Issue by AndrewDryga** [elixir-ecto/ecto#2609]
Currently, this code would silently pass the validation and leave changeset valid:
```
Ecto.Changeset.validate_required(changeset, nil)
```

I believe it's bad because if you use module attributes to keep list of fields it's easy to get `nil` value in the attribute after a minor renaming, eg:
```
@settings_types %{application_id: :string, page_id: :string, access_token: :string}
@settings_fields Map.keys(@settings_types)
@settings_required_fields @fields
```

`@settings_required_fields` would contain `nil`, because it refers to `@fields` which is now called `@settings_fields`.

**Comment by josevalim** [comment]
Good catch. A `not is_nil` guard should be enough. We may have the same issue in the `cast` function.
